channel2/reconnecting_dialer.go

```diff
@@ -131,12 +127,12 @@ func (dialer *reconnectingDialer) sendHello(impl *reconnectingImpl) error {
 		request.Headers[ConnectionIdHeader] = []byte(impl.connectionId)
 		log.Debugf("adding connectionId header [%s]", impl.connectionId)
 	}
```

Comment:
Tx has reconnect code in it, tx doesn't. So sendHello was called from inside the reconnect logic but the send failed, it would try to reconnect. That would hit the reconnect mutex which was already locked by the original entry into the reconnect code, and this would deadlock